### ISSUE TITLE
CDAP endpoint no longer available if Data Fusion instance updated.

### ISSUE THREAD

**Issue by FVCCarneiro** [GoogleCloudPlatform/terraform-provider-cdap#102]
So consider the scenario where I have a Data Fusion in version 6.4.1 and I wish to re-deploy it as 6.5.0 version via Terraform. In Terraform, this implies simply changing the attribute `version` attribute. After making this change, when performing a `terraform plan` the following error is obtained:

`| ERROR Get "/v3/namespaces" : unsupported protocol scheme""`

To solve this, I am forced to remove all cdap ressources which produce the error using the `terraform state rm` command. Only after this can Terraform update the instance and re-create all the CDAP resources that have to be re-created due to the fact that the Data Fusion instance needs to be destroyed and recreated. 

I believe that this error comes from the fact that there is a dependency between a Data Fusion instance and all its CDAP resources that is not considered by the CDAP provider. One should never have to resort to editing the tfstate file to solve these types of issues. Is there a way to fix this?

**Comment by umairidris** [comment]
Thanks for the additional context.

If we remove the wait healthy module, does it work correctly? i.e. a CDF recreation correctly recreates all cdap resources? If so, then it seems we just need to work with the wait healthy module. Maybe we need to add some triggers in the module to correctly recreate when cdf instance is recreated?

Another best practice could be to separate cdf instance and cdap resources into separate states and apply them separately. If that doesn't work, it may require fixes in the provider existence checks that I mentioned above.

**Comment by jrstarke** [comment]
@umairidris do you know if anyone is actively developing on this provider anymore?

**Comment by umairidris** [comment]
Hi @jrstarke, no this provider is not currently being developed. If you would like to send bug fixes or feature improvements then I am happy to review and merge them on a voluntary basis.

If you would like this provider to be better maintained I would suggest sending feedback to the data fusion team: https://cloud.google.com/data-fusion/docs/support/getting-support.

**Comment by FVCCarneiro** [comment]
Hello,

@umairidris , to get back to this issue that still persists for me, the reason why we added the `${module.datafusion_instance.wait_healthy_service_endpoint}/api/`, is because if we follow the [documentation](https://registry.terraform.io/providers/GoogleCloudPlatform/cdap/latest/docs), we get an error in Terraform because the Data Fusion service endpoint take time to be up and therefore the Terraform apply crashes. Perhaps this issue we encountered is no longer relevant ?

**Comment by umairidris** [comment]
Hi @FVCCarneiro,

My apologies, this completely slipped past me!

If you (or someone) is still dealing with this, the first step would be to see if it is an issue with the wait_healthy module. After the initial creation of the null_resource in the wait_healthy module, the CDF instance URL is saved into the TF state and it does not call the script again to check if the instance is still OK.

Thus, we would need to add a trigger to recheck the state of the instance. Looking at https://registry.terraform.io/providers/hashicorp/null/latest/docs/resources/resource, there are some trigger fields that we could set. I am not sure how to make it check for health every time, but https://ilhicas.com/2019/08/17/Terraform-local-exec-run-always.html has a clever way of doing it (by using timestamp as trigger).

Longer term we should not rely on the wait_healthy module. Waiting for health should be done by the CDF instance itself (e.g. we should be able to configure it as an [option](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/data_fusion_instance#options) on the instance). Please file a ticket with the CDF team to add support for this.

Hope this helps!
